libs/animation/animation.ts

```diff
@@ -1,173 +1,624 @@
 /*
     Animation library for sprites
 */
-//% color="#03AA74" weight=78 icon="\uf021"
+//% color="#03AA74" weight=78 icon="\uf021" block="Animation"
+//% groups='["Animate", "Legacy"]'
 namespace animation {
-    //Handles all the updates
-    let animations: Animation[];
-
-    let animationStateStack: {
-        state: Animation[],
-        scene: scene.Scene
-    }[];
-
-    game.addScenePushHandler(oldScene => {
-        if (animations) {
-            if (!animationStateStack) animationStateStack = [];
-            animationStateStack.push({
-                state: animations,
-                scene: oldScene
-            });
-            animations = undefined;
-        }
-    });
-
-    game.addScenePopHandler(() => {
-        const scene = game.currentScene();
-        animations = undefined;
-        if (animationStateStack && animationStateStack.length) {
-            const nextState = animationStateStack.pop();
-            if (nextState.scene == scene) {
-                animations = nextState.state;
-            } else {
-                animationStateStack.push(nextState);
-            }
-        }
-    });
+    const stateNamespace = "__animation";
 
-    export class Animation {
+    interface AnimationState {
+        animations: SpriteAnimation[];
+    }
 
-        sprites: Sprite[];
-        frames: Image[];
-        index: number;
-        interval: number;
-        action: number;
-        lastTime: number;
+    export class Point {
+        public x: number;
+        public y: number;
+        
+        constructor(x: number, y: number) {
+            this.x = x;
+            this.y = y;
+        }
+    }
+
+    //% fixedInstances blockId=animation_path block="path %pathString"
+    export class PathPreset {
+        constructor(public pathString: string) {
+        }
+    }
 
-        constructor(action: number, interval: number) {
-            this.interval = interval;
-            this.index = -1;
-            this.action = action;
-            this.frames = [];
-            this.sprites = [];
-            this.lastTime = control.millis();
+    export class Path {
+        protected nodes: PathNode[];
+        protected lastNode: number; // The index of the last node to fire
 
-            this._init();
+        constructor() {
+            this.nodes = [];
+            this.lastNode = -1;
         }
 
-        _init() {
-            if (!animations) {
-                animations = [];
-                game.eventContext().registerFrameHandler(scene.ANIMATION_UPDATE_PRIORITY, () => {
-                    animations.forEach(anim => anim.update());
-                });
+        private static generateNode(p0: Point, command: string, args: number[], metadata: { pathStart: Point, lastNode: PathNode }): PathNode {
```

Comment:
don't think it actually matters, but would it be reasonable to switch ``metaData`` to be a tuple instead and then you can just destructure it?

```suggestion
        private static generateNode(p0: Point, command: string, args: number[], metadata: [Point, PathNode]): PathNode {
            const pathStart, lastNode = metadata;
```

Would need to change bits in the rest of the function. Just a quick thought, don't know if it's actually a good idea at all